python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1183,195 @@ def _get_or_compile(
+
+        Returns:
+            A graph that each node is a DAGOperationGraphNode.
+        """
+        assert self.idx_to_task
+
+        graph: Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]] = {}
+
+        for _, operation_nodes_list in actor_to_operation_nodes.items():
+            prev_compute_node = None
+            for operation_nodes in operation_nodes_list:
+                idx = operation_nodes[0].idx
+                read_node, compute_node, write_node = (
+                    operation_nodes[0],
+                    operation_nodes[1],
+                    operation_nodes[2],
+                )
+                # Add edges from READ to COMPUTE, and from COMPUTE to WRITE, which
+                # belong to the same task.
+                read_node.add_edge(compute_node)
+                compute_node.add_edge(write_node)
+                # Add an edge from COMPUTE with `bind_index` i to COMPUTE with
+                # `bind_index` i+1 if they belong to the same actor.
+                if prev_compute_node is not None:
+                    prev_compute_node.add_edge(compute_node)
+                prev_compute_node = compute_node
+                assert idx not in graph
+                graph[idx] = {
+                    DAGNodeOperationType.READ: read_node,
+                    DAGNodeOperationType.COMPUTE: compute_node,
+                    DAGNodeOperationType.WRITE: write_node,
+                }
+
+        from ray.dag import (
+            ClassMethodNode,
+            MultiOutputNode,
+        )
+
+        # Add an edge from WRITE of the writer task to READ of the reader task.
+        for idx, task in self.idx_to_task.items():
+            if not isinstance(task.dag_node, ClassMethodNode):
+                continue
+            for downstream_idx in task.downstream_node_idxs:
+                downstream_dag_node = self.idx_to_task[downstream_idx].dag_node
+                if isinstance(downstream_dag_node, MultiOutputNode):
+                    continue
+                graph[idx][DAGNodeOperationType.WRITE].add_edge(
+                    graph[downstream_idx][DAGNodeOperationType.READ]
+                )
+        return graph
+
+    def _build_execution_schedule(self):
+        """
+        Generate an execution schedule for each actor. The schedule is a list of
+        DAGNodeOperation.
+
+        Step 1: Generate a DAG node operation graph. Refer to the functions
+        `_generate_dag_operation_graph_node` and `_build_dag_node_operation_graph`
+        for more details.
+
+        Step 2: Topological sort
+
+        It is possible to have multiple DAGOperationGraphNodes with zero in-degree.
+        Refer to the function `_select_next_nodes` for the logic of selecting nodes.
+
+        Then, put the selected nodes into the corresponding actors' schedules.
+
+        The goal of the above rules is to build a schedule with fewer bubbles. The
```

Comment:
Question: which rules would reduce bubbles? Executing multiple operations in parallem if they all have in_degree=0?

Reply:
Good question. I think the comment here is a bit confusing because the baseline is unclear. The baselines here are the other two solutions we proposed internally. `in_degree=0` primarily contributes to avoiding deadlock. Selecting the one with the smaller `bind_index` is the primary factor that reduces bubbles.

Because the baselines are unclear in this context, I will remove the comment.

Reply:
Removed https://github.com/ray-project/ray/pull/46911/commits/5c3750ed58716da39fb2b93e17b9af225c9a0fee